### PR TITLE
refactor(iroh)!: Extract net and node rpc

### PR DESCRIPTION
**Pull request by rklaehn**
## Description

Introduce a new crate iroh-node-util.

This contains generic utilities that are needed to build an iroh node, as well as the rpc protocol and rpc client to remote control an arbitrary iroh-net endpoint (add/remove addresses etc.) as well as to control the node as a whole (shutdown/stats/status).

## Breaking Changes

- module iroh::client::net is now a reexport of iroh_node_util::rpc::client::net
- module iroh::client::node is now a reexport of iroh_node_util::rpc::client::node
- fn iroh::client::Iroh::net returns a net::Client instead of a &net::Client
- fn iroh::client::Iroh::node returns a node::Client instead of a &node::Client

## Notes & open questions

Note: I think the rpc protocol and the client needs some refactoring. It isn't always clear to me what is node and what is net. But I think it might be best to do this in a subsequent PR and keep this one as a pure move things around PR for better separation.

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [ ] Tests if relevant.
- [ ] All breaking changes documented.

### CODE REVIEW

File: iroh-node-util/src/rpc/client.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+use anyhow::Result;
+use futures_lite::{Stream, StreamExt};
+
+pub mod net;
+pub mod node;
+
+fn flatten<T, E1, E2>(
```
Comment (dignifiedquire) [inline iroh-node-util/src/rpc/client.rs:7]: maybe add this to quic-rpc? its used everywhere 🤣

File: iroh/src/client.rs [hunk 2]
```diff
@@ -80,42 +77,30 @@ impl Iroh {
...
     }
 
     /// Shuts down the node.
     ///
     /// If `force` is true, the node will be shut down instantly without
     /// waiting for things to stop gracefully.
     pub async fn shutdown(&self, force: bool) -> Result<()> {
-        self.rpc.rpc(ShutdownRequest { force }).await?;
-        Ok(())
+        self.node().shutdown(force).await
```
Comment (dignifiedquire) [inline iroh/src/client.rs:94]: we should drop these helper methods, now that they are on `.node()`

File: iroh-node-util/src/rpc/server.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        }
+    }
+
+    async fn handle_net_request<C: ChannelTypes<RpcService>>(
+        self,
+        msg: net::Request,
+        chan: RpcChannel<RpcService, C>,
+    ) -> Result<(), RpcServerError<C>> {
+        use net::Request::*;
+        debug!("handling node request: {msg}");
```
Comment (dignifiedquire) [inline iroh-node-util/src/rpc/server.rs:71]: net, not node

**Comment by dignifiedquire** [comment]
I like the structure, but I am not clear that we really need this in its own repo, or if we just leave it for the moment in `iroh`

File: iroh-node-util/src/rpc/client.rs (hunk 1 above)
Comment (rklaehn) [reply iroh-node-util/src/rpc/client.rs:7]: Where is this used? In the context of streaming responses, right? I can maybe add it somewhere in client as an util...

File: iroh-node-util/src/rpc/server.rs (hunk 3 above)
Comment (rklaehn) [reply iroh-node-util/src/rpc/server.rs:71]: it's confusing...

**Comment by rklaehn** [comment]
> I like the structure, but I am not clear that we really need this in its own repo, or if we just leave it for the moment in iroh

This would help with the insane amount of boilerplate you need when you want to write an iroh node for a specific set of protocols, like e.g. I have to do in the iroh-docs tests here: https://github.com/n0-computer/iroh-docs/blob/main/tests/util.rs#L36

Ideally it should be quite easy to write an iroh node for a specific set of protocols, complete with a bin and a cli that talks to the bin via rpc.

File: iroh-node-util/src/rpc/client.rs (hunk 1 above)
Comment (rklaehn) [reply iroh-node-util/src/rpc/client.rs:7]: Not super urgent, but I am adding it to quic-rpc: https://github.com/n0-computer/quic-rpc/pull/120

File: Cargo.toml [hunk 4]
```diff
@@ -11,7 +11,7 @@ members = [
   "iroh-relay",
   "iroh-router",
   "net-tools/netwatch",
-  "net-tools/portmapper",
+  "net-tools/portmapper", "iroh-node-util",
```
Comment (dignifiedquire) [inline Cargo.toml:14]: formatting

File: iroh-node-util/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "iroh-node-util"
+license = "MIT OR Apache-2.0"
+version = "0.1.0"
+edition = "2021"
+authors = ["n0 team"]
+repository = "https://github.com/n0-computer/iroh"
+keywords = ["quic", "networking", "holepunching", "p2p"]
+
```
Comment (dignifiedquire) [inline iroh-node-util/Cargo.toml:9]: missing some fields we use on other cargo.tomls

File: iroh-node-util/src/rpc/client/net.rs [hunk 6]
```diff
@@ -13,116 +11,84 @@ use std::net::SocketAddr;
...
-#[derive(Debug, Clone, RefCast)]
+#[derive(Debug, Clone)]
 #[repr(transparent)]
 pub struct Client {
-    pub(super) rpc: RpcClient,
+    pub(super) rpc: RpcClient<RpcService>,
 }
 
 impl Client {
+    pub fn new(rpc: RpcClient<RpcService>) -> Self {
```
Comment (dignifiedquire) [inline iroh-node-util/src/rpc/client/net.rs:42]: docs?

File: iroh-node-util/src/rpc/client/node.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+use crate::rpc::proto::{node::*, RpcService};
+
+#[derive(Debug, Clone)]
+#[repr(transparent)]
+pub struct Client {
+    pub(super) rpc: RpcClient<RpcService>,
+}
+
+impl Client {
+    pub fn new(rpc: RpcClient<RpcService>) -> Self {
```
Comment (dignifiedquire) [inline iroh-node-util/src/rpc/client/node.rs:16]: docs

File: iroh-node-util/Cargo.toml (hunk 5 above)
Comment (rklaehn) [reply iroh-node-util/Cargo.toml:11]: added description and readme. not sure what else to add.

File: iroh-node-util/README.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# iroh-base
```
Comment (dignifiedquire) [inline iroh-node-util/README.md:1]: wrong name

File: iroh/Cargo.toml [hunk 9]
```diff
@@ -29,6 +29,7 @@ iroh-docs = { version = "0.28.0", features = ["rpc"] }
 iroh-io = { version = "0.6.0", features = ["stats"] }
 iroh-metrics = { version = "0.28.0", optional = true }
 iroh-net = { version = "0.28.1", features = ["discovery-local-network"] }
+iroh-node-util = { path = "../iroh-node-util" }
```
Comment (dignifiedquire) [inline iroh/Cargo.toml:32]: please add a version

File: iroh-node-util/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "iroh-node-util"
+description = "Utilities to build binaries containing an iroh-net endpoint"
+readme = "README.md"
+license = "MIT OR Apache-2.0"
+version = "0.1.0"
```
Comment (dignifiedquire) [inline iroh-node-util/Cargo.toml:6]: please make this `0.28`

**Comment by dignifiedquire** [comment]
please document the breaking changes, otherwise I left some last comments and then this should be good to go

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**